File: python/toy_robot.py
import os


class ToyRobot:
    DIRECTIONS = ["NORTH", "EAST", "SOUTH", "WEST"]
    ICONS = {
        "NORTH": "↑",
        "EAST": "→",
        "SOUTH": "↓",
        "WEST": "←"
    }

    def __init__(self, size=5):
        self.size = size
        self.x = None
        self.y = None
        self.facing = None
        self.placed = False

    def place(self, x, y, facing):
        if self._is_valid_position(x, y) and facing in self.DIRECTIONS:
            self.x, self.y, self.facing = x, y, facing
            self.placed = True
            print(f"Placed at {x},{y},{facing}")
            self.draw_table()
            return True
        else:
            print("Invalid PLACE command.")
            return False

    def move(self):
        if not self.placed:
            print("Robot not yet placed.")
            return False

        new_x, new_y = self.x, self.y
        if self.facing == "NORTH":
            new_y += 1
        elif self.facing == "SOUTH":
            new_y -= 1
        elif self.facing == "EAST":
            new_x -= 1
        elif self.facing == "WEST":
            new_x += 1

        if self._is_valid_position(new_x, new_y):
            self.x, self.y = new_x, new_y
            print(f"Moved to {self.x},{self.y}")
            self.draw_table()
            return True
        else:
            print("Move ignored to prevent falling off the table.")
            return False

    def left(self):
        if not self.placed:
            print("Robot not yet placed.")
            return False
        idx = (self.DIRECTIONS.index(self.facing) - 1) % 4
        self.facing = self.DIRECTIONS[idx]
        print(f"Turned left. Now facing {self.facing}")
        self.draw_table()
        return True

    def right(self):
        if not self.placed:
            print("Robot not yet placed.")
            return False
        idx = (self.DIRECTIONS.index(self.facing) + 1) % 4
        self.facing = self.DIRECTIONS[idx]
        print(f"Turned right. Now facing {self.facing}")
        self.draw_table()
        return True

    def report(self):
        if not self.placed:
            print("Robot not yet placed.")
            return False
        self.draw_table()
        print(f"Output: {self.x},{self.y},{self.facing}")
        return True

    def draw_table(self):
        os.system('cls' if os.name == 'nt' else 'clear')
        print("\nToy Robot Simulator\n")

        for y in reversed(range(1, self.size + 1)):
            row = []
            for x in reversed(range(1, self.size + 1)):
                if self.placed and x == self.x and y == self.y:
                    row.append(self.ICONS[self.facing])
                else:
                    row.append("·")
            print(" ".join(row) + f"  | {y}")
        print("5 4 3 2 1\n")
        print("Legend: ↑=NORTH, →=EAST, ↓=SOUTH, ←=WEST\n")

    def _is_valid_position(self, x, y):
        if 1 <= x <= self.size and 1 <= y <= self.size:
            return True
        else:
            return False

File: typescript/toy_robot.ts
export class ToyRobot {
  private size: number;
  private x: number | null = null;
  private y: number | null = null;
  private facing: string | null = null;
  private placed = false;

  private readonly DIRECTIONS = ["NORTH", "EAST", "SOUTH", "WEST"] as const;
  private readonly ICONS: any = {
    NORTH: "↑",
    EAST: "→",
    SOUTH: "↓",
    WEST: "←",
  };

  constructor(size = 5) {
    this.size = size;
  }

  public place(x: number, y: number, facing: string) {
    if (this.isValidPosition(x, y) && this.DIRECTIONS.includes(facing as any)) {
      this.x = x;
      this.y = y;
      this.facing = facing;
      this.placed = true;
      console.log(`Placed at ${x},${y},${facing}`);
      this.drawTable();
    } else {
      console.log("Invalid PLACE command.");
    }
  }

  public move() {
    if (!this.placed || this.x === null || this.y === null || this.facing === null) {
      console.log("Robot not yet placed.");
      return;
    }

    let newX = this.x;
    let newY = this.y;

    switch (this.facing) {
      case "NORTH":
        newY += 1;
        break;
      case "SOUTH":
        newY -= 1;
        break;
      case "EAST":
        newX -= 1;
        break;
      case "WEST":
        newX += 1;
        break;
    }

    if (this.isValidPosition(newX, newY)) {
      this.x = newX;
      this.y = newY;
      console.log(`Moved to ${this.x},${this.y}`);
      this.drawTable();
    } else {
      console.log("Move ignored to prevent falling off the table.");
    }
  }

  public left() {
    if (!this.placed || this.facing === null) {
      console.log("Robot not yet placed.");
      return;
    }
    const idx = (this.DIRECTIONS.indexOf(this.facing as any) - 1 + 4) % 4;
    this.facing = this.DIRECTIONS[idx];
    console.log(`Turned left. Now facing ${this.facing}`);
    this.drawTable();
  }

  public right() {
    if (!this.placed || this.facing === null) {
      console.log("Robot not yet placed.");
      return;
    }
    const idx = (this.DIRECTIONS.indexOf(this.facing as any) + 1) % 4;
    this.facing = this.DIRECTIONS[idx];
    console.log(`Turned right. Now facing ${this.facing}`);
    this.drawTable();
  }

  public report() {
    if (!this.placed || this.x === null || this.y === null || this.facing === null) {
      console.log("Robot not yet placed.");
      return;
    }
    this.drawTable();
    console.log(`Output: ${this.x},${this.y},${this.facing}`);
  }

  private drawTable() {
    console.clear();
    console.log("\nToy Robot Simulator\n");

    for (let y = this.size; y >= 1; y--) {
      const row: string[] = [];
      for (let x = this.size; x >= 1; x--) {
        if (this.placed && x === this.x && y === this.y && this.facing) {
          row.push(this.ICONS[this.facing]);
        } else {
          row.push("·");
        }
      }
      console.log(`${row.join(" ")}  | ${y}`);
    }
    console.log("5 4 3 2 1\n");
    console.log("Legend: ↑=NORTH, →=EAST, ↓=SOUTH, ←=WEST\n");
  }

  private isValidPosition(x: number, y: number): boolean {
    return x >= 1 && x <= this.size && y >= 1 && y <= this.size;
  }
}
